File: algorithm/leetcodeCppTmp/top-k-frequent-elements.cpp
// 347. Top K Frequent Elements
// Difficulty: Medium

// Given a non-empty array of integers, return the k most frequent elements.

// For example,
// Given [1,1,1,2,2,3] and k = 2, return [1,2].

// Note: 
// You may assume k is always valid, 1 ≤ k ≤ number of unique elements.
// Your algorithm time complexity must be better than O(n log n), where n is the array size.

// Time:  O(nlogk)
// Space: O(n)
// Heap solution.
class Solution2 {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> counts;
        for (int i = 0; i < nums.size(); ++i) {
            ++counts[nums[i]];
        }
        priority_queue<pair<int, int>> heap;
        for (auto it = counts.begin(); it != counts.end(); ++it) {
            heap.emplace(-(it->second), it->first);
            if (heap.size() == k + 1) {
                heap.pop();
            }
        }
        vector<int> result;
        while (!heap.empty()) {
            result.emplace_back(heap.top().second);
            heap.pop();
        }
        reverse(result.begin(), result.end());
        return result;
    }
};

File: data-structure-and-algorithm/javascript/leetcode/top-k-frequent-elements.js
/*
Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

 

Example 1:

Input: nums = [1,1,1,2,2,3], k = 2
Output: [1,2]
Example 2:

Input: nums = [1], k = 1
Output: [1]
 

Constraints:

1 <= nums.length <= 105
k is in the range [1, the number of unique elements in the array].
It is guaranteed that the answer is unique.
 

Follow up: Your algorithm's time complexity must be better than O(n log n), where n is the array's size.
*/

function topKFrequent(nums, k) {
    const count = new Map();
    for (const n of nums) {
        count.set(n, (count.has(n) ? count.get(n) : 0) + 1);
    }
    const bucket = [...Array(nums.length + 1)].map(() => [])
    for (const [n, c] of count.entries()) {
        bucket[c].push(n);
    }
    const result = [];
    for (let i = bucket.length - 1; i >= 0; --i) {
        for (let j = 0; j < bucket[i].length; ++j) {
            result.push(bucket[i][j]);
            if (result.length === k) {
                return result;
            }
        }
    }
    return result;
}